### ISSUE TITLE
evil eval from download -> got -> create-error-class fails badly

### ISSUE THREAD

**Issue by hedefalk** [ensime/ensime-node#8]
Got errors from got using eval that fails when run on ui thread in Atom which will happen if require on top level.

**Comment by hedefalk** [comment]
https://github.com/AtomLinter/linter-eslint/issues/210

**Comment by hedefalk** [comment]
This doesn't seem to fail in Atom when running anymore, but it fails on ci:

https://travis-ci.org/ensime/ensime-atom/jobs/120055214:

```
ensime-server-update
  it should be able to download coursier
    EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".

      EvalError: Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'".
      at createErrorClass (/home/travis/build/ensime/ensime-atom/node_modules/download/node_modules/got/node_modules/create-error-class/index.js:18:49)
      at Object.<anonymous> (/home/travis/build/ensime/ensime-atom/node_modules/download/node_modules/got/index.js:363:20)
      at Module._compile (/home/travis/atom/usr/share/atom/resources/app.asar/src/native-compile-cache.js:103:30)
      at Object.defineProperty.value [as .js] (/home/travis/atom/usr/share/atom/resources/app.asar/src/compile-cache.js:208:21)
      at Module.load (/home/travis/atom/usr/share/atom/resources/app.asar/node_modules/coffee-script/lib/coffee-script/register.js:45:36)
      at Function.Module._load (module.js:310:12)
      at Module.require (module.js:365:17)
      at require (/home/travis/atom/usr/share/atom/resources/app.asar/src/native-compile-cache.js:50:27)
      at Object.<anonymous> (/home/travis/build/ensime/ensime-atom/node_modules/download/index.js:9:11)
      at Module._compile (/home/travis/atom/usr/share/atom/resources/app.asar/src/native-compile-cache.js:103:30)
      at Object.defineProperty.value [as .js] (/home/travis/atom/usr/share/atom/resources/app.asar/src/compile-cache.js:208:21)
      at Module.load (/home/travis/atom/usr/share/atom/resources/app.asar/node_modules/coffee-script/lib/coffee-script/register.js:45:36)
      at Function.Module._load (module.js:310:12)
      at Module.require (module.js:365:17)
      at require (/home/travis/atom/usr/share/atom/resources/app.asar/src/native-compile-cache.js:50:27)
      at /home/travis/build/ensime/ensime-atom/node_modules/ensime-client/lib/ensime-server-update-coursier.js:70:20
      at [object Object].<anonymous> (/home/travis/build/ensime/ensime-atom/spec-integration/ensime-server-update-coursier-spec.coffee:28:5)
```

My uneducated guess is that it worked in my Atom when I delayed it to run on the chromium thread (https://github.com/ensime/ensime-node/commit/abd4ece521b184adfdb3a36f7dcd20822d37f889), but not in normal node which might run the init of all scripts? 

Anyway, I think we should really use something else that doesn't use `eval`.

**Comment by hedefalk** [comment]
The eval has been removed at the end dep:

https://github.com/floatdrop/create-error-class/commit/c899536f5e1432a1b8846493fe8047978f5c5b00

**Comment by hedefalk** [comment]
Download is 4 months since last release and uses got ^5.0.0:

https://github.com/kevva/download/blob/master/package.json#L33

Got has updated long ago: https://github.com/sindresorhus/got/commits but obviously not in 5.x.x series. So download is the problem.

**Comment by hedefalk** [comment]
@jeckhart Do you know if we have any options to download in case we don't get any luck there?

**Comment by hedefalk** [comment]
@steelbrain Quicky question about eval: Is it possible to workaround in any way? You had similar problems with linter? https://github.com/AtomLinter/linter-eslint/issues/210

**Comment by hedefalk** [comment]
Oh, my fix did nothing but delay the error until download was required. I thought that maybe there was something about ui thread vs node or something, but this simply doesn't work in Atom. Download is broken in Atom. Kindof gotta revert this change or find a replacement…

**Comment by steelbrain** [comment]
@hedefalk There are several workarounds for this, one is the [`loophole`](https://www.npmjs.com/package/loophole) npm package, the other is to spawn a node worker using the Atom executable as the electron runtime. We chose the second option

**Comment by hedefalk** [comment]
@steelbrain Cool, thanks!

I actually started out doing npm shrinkwrap to fix the version of the transitive dependency to a later one without eval. But I misunderstood shrinkwrap. Basically I did like https://github.com/npm/npm/issues/7108#issue-53955766, but this failed on travis for some versions of node. I then tried it locally with those version and yes, I guess npm has changed on this. I thought I could just put a subset of stuff in shrinkwrap, but then stuff wasn't installed for older versions of npm:

https://travis-ci.org/hedefalk/ensime-node/builds/120089602

I guess I could just do a postinstall like suggested here: 
https://github.com/npm/npm/issues/7108#issuecomment-69507982

```
find node_modules -type d -name got | xargs rm -fr
npm install got@6.2.0
```

I am a good citizen though and put a PR on the upstream package: https://github.com/kevva/download/pull/95

Seems a bit abandon though :)
